### PR TITLE
append new pdf pages via menu

### PR DESCRIPTION
**Pull request by rolandlo**
The idea of this PR is to provide a convenient option to append all new pdf pages to a Xournal++ document after the pdf has been updated externally with new pages. This is already possible by adding one page after the other, but that is not so practical if lots of pages have been added, see issue #1588.

Unfortunately there is bug #2144 which needs to be fixed. I have only made a temporary fix, which softens the problem, but does not quite solve it. With the temporary fix you won't have any freezes, but the page number is not updated if you change to the very first page of your document. This would have to be addressed before the PR can be merged.

### CODE REVIEW

**Review by LittleHuba — CHANGES_REQUESTED** [review]
Looks good to me so far. The warnings should definitely be provided as a dialog though. Most users do not see such messages posted to stderr.

**Comment by rolandlo** [comment]
@LittleHuba The errors are now reported in XojMsgBox'es. Do you have an idea about the freezing bug #2144 (which I "fixed" at the cost of pages not updating when scrolling to page 1)? It was introduced after you reintroduced the "select page after scrolling" algorithm. I haven't figured out yet, whether we should introduce a condition to break the infinite loop or whether there is something wrong with the algorithm to compute the most visible page (when there is a pdf background).

**Comment by LittleHuba** [comment]
That is a taff problem. I'll look into that issue tonight. Most likely we have to separate the algorithm into two different ones.

File: src/control/Control.cpp
Comment (Febbe) [inline src/control/Control.cpp:1252]: All of those return values are size_t, so don't cast to int implicitly without bound checks.

```suggestion
    auto pageCount = this->doc->getPageCount();
    // find last page with pdf background and get its pdf page number
    auto currentPdfPageCount = [&]() {
        for (size_t i = pageCount; i != 0; --i) {
            if (auto page = doc->getPage(i - 1); page && page->getBackgroundType().isPdfPage()) {
                return page->getPdfPageNr() + 1;
            }
        }
        return 0ULL;
    }();
```

File: src/control/Control.cpp
Comment (Febbe) [inline src/control/Control.cpp:1257]: I think in this cornercase should we should add the new pages. Imagine a document with all PDF pages deleted, and you add new pages to the PDF, now all the pdfs should be added. Am I right?

```suggestion
```

File: src/control/Control.cpp
Comment (Febbe) [inline src/control/Control.cpp:1262]: Also use auto or size_t.
```suggestion
    auto pdfPageCount = this->doc->getPdfPageCount();
    auto insertCount = pdfPageCount - currentPdfPageCount;

    if (insertCount == 0) {
```

File: src/control/Control.cpp
Comment (Febbe) [inline src/control/Control.cpp:1275]:
```suggestion
    for (size_t i = 0; i != insertCount; ++i) {

        doc->lock();
        XojPdfPageSPtr pdf = doc->getPdfPage(currentPdfPageCount + i);
        doc->unlock();

        if (pdf) {
            auto newPage = std::make_shared<XojPage>(pdf->getWidth(), pdf->getHeight());
            newPage->setBackgroundPdfPageNr(currentPdfPageCount + i);
            insertPage(newPage, pageCount + i);
```

**Review by Febbe — CHANGES_REQUESTED** [review]

File: src/gui/Layout.cpp
Comment (rolandlo) [reply src/gui/Layout.cpp:112]: Yes, we don't want that. It should be fixed. That's why I added the "Do not merge label". 
What do you mean by "empty pdf pages"? Do you mean pdf pages with no actual content (just white background)? I never have empty pdf pages in my documents and I can't think of documents that have, unless the creator wanted them to be there. In any case, deleting empty pages at the end is a lot easier than adding pdf pages that have been created externally (for instance by adding new content to a LaTeX document or adding content to a Word- or Powerpoint-document and re-exporting to pdf).

File: src/control/Control.cpp
Comment (rolandlo) [reply src/control/Control.cpp:1252]: This syntax is new to me. What does the `[&](){ .... }();` stuff do?

File: src/control/Control.cpp
Comment (rolandlo) [reply src/control/Control.cpp:1257]: Ok, I am fine with adding all pdf pages in that case. If all pdf pages are deleted and you save/reload the document there won't be any pdf pages to add, though. It only occurs if you remove the last pdf page in the document and don't save.

File: src/control/Control.cpp
Comment (Febbe) [reply src/control/Control.cpp:1252]: > This syntax is new to me. What does the `[&](){ .... }();` stuff do?

Oh that's a lamda expression. It is part of the standard since c++11.
Basically I inlined a unnamed function here, returning the value for "currentPdfPageCount" and with the prceding `()` I directly executed it.
With that no predeclared variable is necessary and the returnvalue of the lambda function is directly executed.

https://en.cppreference.com/w/cpp/language/lambda

File: src/gui/Layout.cpp
Comment (Febbe) [reply src/gui/Layout.cpp:112]: I misunderstood #2211

File: src/control/Control.cpp
Comment (Febbe) [inline src/control/Control.cpp:1250]: This is only correct on 32bit or Linux machines.
`return ULL;` was already correct, since we don't support 32 bit.
But you can support all machines / compiler:
```suggestion
        return size_t{0U};
```

**Comment by Febbe** [comment]
LGTM after a fixup-rebase.

File: src/control/Control.cpp
Comment (rolandlo) [reply src/control/Control.cpp]: Thanks for the fix. It didn't compile with `return 0ULL` on my (64-bit Linux) machine (gcc 9.2.1). That's why I changed it. I am glad you checked it once more. With `size_t{0U}` it works.

**Comment by rolandlo** [comment]
After bug #2144 has been fixed, this PR looks ready for merging to me. I have removed the workaround for bug #2144 in the code, so it just works without problems.

**Pull request merged**
